Make sure to revalidate instances when mixins are invalidated

dfreedm opened the pull request:
Fixes #21

dfreedm committed:
Make sure to revalidate instances when mixins are invalidated

Fixes #21

dfreedm committed:
Instead of tracking instances, use a timeout

No risk of memory leaks

dfreedm committed:
use microtask timing instead of a timeout

dfreedm committed:
compile makeElement code for tests

dfreedm committed:
polyfill Object.assign

TomK commented:
this works for me 👍

dfreedm committed:
Merge branch 'master' into mixin-ordering

sorvell commented on src/ShadyCSS.js:
```diff
@@ -150,17 +152,21 @@ export let ShadyCSS = {
     if (!styleInfo) {
       styleInfo = this._prepareHost(host);
     }
+    assign(styleInfo.overrideStyleProperties, overrideProps);
```
remove

sorvell commented on src/ShadyCSS.js:
```diff
@@ -180,18 +186,25 @@ export let ShadyCSS = {
+        if (!nativeShadow) {
+          window.ShadyDOM.flush();
+        }
+        this._applyToDescendants(c.assignedNodes().filter(this._isElementNode));
```
Let's add an inline comment about why this needs to be here in this exact case.

dfreedm committed:
Always recurse when elements boot, but prevent unnecessary style recalculations

Elements will always recurse style updates down the tree, and it is up
to those descendant elements to decide if they need to recalculate
styles.

The four scenarios where they will recalculate styles are:
1. The @apply shim has marked the template invalid
2. The user has supplied style overrides
3. The element has never applied styles
4. The element has been marked stale (removed from the DOM and
reinserted).

dfreedm committed:
remove sauce_connect addon

sorvell commented on src/ShadyCSS.js:
```diff
@@ -146,52 +149,74 @@ export let ShadyCSS = {
+      c = children[i];
       if (c.shadowRoot) {
         this.applyStyle(c);
+      } else if (nativeCssVariables && c.localName === 'slot') {
```
Let's consider not doing this as it's a perf hit (yes, that means the issue still exists).

sorvell commented on src/ShadyCSS.js:
```diff
@@ -146,52 +149,74 @@ export let ShadyCSS = {
-          let style = host.shadowRoot.querySelector('style');
-          style.textContent = StyleTransformer.elementStyles(host, styleInfo.styleRules);
+    let template = templateMap[is];
+    if (this._forceRecalc || styleInfo.shouldRecalculate || template._applyShimInvalid) {
```
let's also bail out of this function early if we detect the element does not use or produce `apply` rules and native properties are in use.
